source/Calamari.Aws/Kubernetes/Discovery/AwsKubernetesDiscoverer.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Amazon;
+using Amazon.EKS;
+using Amazon.EKS.Model;
+using Calamari.Common.Features.Discovery;
+using Calamari.Common.Plumbing.Logging;
+using Calamari.Common.Plumbing.Variables;
+using Calamari.Kubernetes.Aws;
+using Newtonsoft.Json;
+using Octopus.CoreUtilities.Extensions;
+
+namespace Calamari.Aws.Kubernetes.Discovery
+{
+    public class AwsKubernetesDiscoverer : IKubernetesDiscoverer
+    {
+        readonly ILog log;
+
+        public AwsKubernetesDiscoverer(ILog log)
+        {
+            this.log = log;
+        }
+
+        public string Name => "Aws";
+        
+        public IEnumerable<KubernetesCluster> DiscoverClusters(string contextJson)
+        {
+            if (!TryGetDiscoveryContext(contextJson, out var discoveryContext))
+                yield break;
+
+            var authenticationDetails = discoveryContext.Authentication;
+            if (authenticationDetails == null)
+            {
+                Log.Warn("Target Discovery Context is in the wrong format: Unable to serialise authentication details");
```

Comment:
It's deserialization that has failed here, not serialization.
Maybe just "read" or "access" is a friendlier message here.
Since these are user facing messages we should make it clear that this is a problem with the application, not a user error.

@rhysparry Are there any guidelines on consistent use of language to inform users of server/application errors which the user cannot do anything about vs errors users can take corrective action for?

Reply:
The [Logging](https://github.com/OctopusDeploy/Architecture/blob/main/Steps/Logging.md) document in the Architecture repo is the one I know about.

I think it's worth considering what a user is expected to do with the information. The "Build succeeded with warnings" is something we'd like users to be able to ensure they can achieve (users don't want to have to inspect the logs to see warnings they can't do anything about).

Reply:
Thanks for the link @rhysparry Good to be reminded of those principles. I've updated the code to better align with those I think 👍